BXMSDOC-414: Add KIE scanner section to packaging content.

sterobin opened the pull request:
Links:
* [JIRA](https://issues.jboss.org/browse/BXMSDOC-414)
* [DM/PAM doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-414_PAM/#kie-scanner-proc_packaging-deploying)
* [Community doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-414_DROOLS/#kie-scanner-proc_decision-management-architecture)

sterobin commented:
Thanks @winklerm. I didn't include those details because they are apparently not recommended for enterprise use, as Marek C. mentioned in the JIRA: https://issues.jboss.org/browse/BXMSDOC-414?focusedCommentId=13244703&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13244703

@agiertli, please have a look and let me know if you approve. If I don't hear from you by the time I have @mswiderski's approval and peer approval, I'l need to merge this and incorporate your improvements separately.

Thanks!

agiertli commented:
> Thanks @winklerm. I didn't include those details because they are apparently not recommended for official enterprise documentation, as Marek C. mentioned in the JIRA: https://issues.jboss.org/browse/BXMSDOC-414?focusedCommentId=13244703&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13244703 That is my understanding as well. Either way, I agree with Marek W. that those are details about how Maven works and if there's any doubt about whether or not it's recommended (sounds like there's doubt), then I think best to be safe and not go into it.
> 
> @agiertli, please have a look and let me know if you approve. If I don't hear from you by the time I have @mswiderski's approval and peer approval, I'l need to merge this and incorporate your improvements separately.
> 
> Thanks!

Hi @sterobin 

you said you didn't document SNAPSHOT because they are not suitable for enterprise use, yet KieScanner only works against non-static version and you are using SNAPSHOT in the example itself. So I am quite confused about this line of reasoning.  You also said KieScanner is suitable only for dev purposes, which I am not really convinced about - can you please explain why do you think so?

@winklerm It's a big assumption to make that all the KieScanner users will have that level of knowledge of the maven internal details - it's up to our documentation to provide that level of detail and explanation. It personally took me quite a while to understand what's not expected and expected behaviour when it comes to KieScaner,  it's not something apparent I think. That's why I thought it should be part of the docs.

sterobin commented:
Thanks @nmirasch.

Spoke with @agiertli in chat and clarified. Updated with version type information but agreed to excluded diagnostic steps, which are too outdated and confusing at this point.

mswiderski commented on doc-content/shared-kie-docs/src/main/asciidoc/KIE/BuildDeployUtilizeAndRun/kie-scanner-proc.adoc:
```diff
@@ -0,0 +1,82 @@
+
+* `x.x-SNAPSHOT`: (For development environments only) If you add the `-SNAPSHOT` suffix to your project version, the KIE scanner retrieves the latest build of the specified artifact version. Example GAV settings: `com.sample:my-app:1.0-SNAPSHOT`. If you change the project version number, such as increasing to `1.0.1-SNAPSHOT`, then you must also update the version in the GAV definition in your KI ...
+* `LATEST`: If you set the project version to `LATEST`, the KIE scanner retrieves the latest version of the specified artifact, regardless of the specified version. Example GAV settings: `com.sample:my-app:LATEST`.
+* `RELEASE`: If you set the project version to `RELEASE`, the KIE scanner retrieves the latest non-snapshot version of the specified artifact. Example GAV settings: `com.sample:my-app:RELEASE`.
```
@mariofusco do we still support (or recommend) LATEST and RELEASE pseudo versions?

sterobin replied:
Thanks @mswiderski. Hahaha! @agiertli, this is what I was saying above about why my assumption was to not document those LATEST and RELEASE options. We'll see what @mariofusco says. Please feel free to give your opinion on it.

sterobin replied:
@mariofusco, @mswiderski, after speaking with @agiertli, I realized I misunderstood his initial feedback above. I thought he was suggesting to incorporate the LATEST and RELEASE info from the KB article as well, as @winklerm pointed out, but turns out he just meant to explain SNAPSHOT vs static versions. So if there's any doubt about this, I'm good to remove it and emphasize that KIE scanner is therefore for development environments only, with SNAPSHOT project versions, and not recommended for production environments.

Let me know what you think.

sterobin replied:
From @mariofusco in chat:

> +1, the fact is that now maven doesn't support anymore LATEST or RELEASE versions bcause they led to not reproducible build and then we should discourage their use as well so ok to avoid mentioning them at all

Removing those and updating now and will ping for final confirmation shortly.

sterobin commented:
@mswiderski, @agiertli, @winklerm, I've removed the content on LATEST and RELEASE, and rewritten the intro portion and the code example description to emphasize that KIE scanner is only for SNAPSHOT projects in development environments, not intended for production environments, doesn't work with static versions, etc.

Use the same doc preview in the description above, but **refresh the doc preview page once opened to get the latest**.

Let me know what you think. Thanks!

winklerm commented:
@sterobin the changes look fine to me, thanks!

sterobin commented:
Confirmed with @agiertli in chat that he is fine with the changes. He expressed some questions about the prescription of using this in dev vs. prod, but agreed that it was likely a difference of opinion and is fine leaving the content now as is.

csherrar commented on doc-content/shared-kie-docs/src/main/asciidoc/KIE/BuildDeployUtilizeAndRun/kie-scanner-proc.adoc:
```diff
@@ -0,0 +1,83 @@
+
+[IMPORTANT]
+====
+For production environments, do not use a KIE scanner or `SNAPSHOT` project versions to avoid accidental or unexpected project updates. The KIE scanner is intended for development environments that use `SNAPSHOT` project versions.
```
There is a mix of plural and singular. Is KIE scanner the name of a service or can it be multiple? If it is, do you think the indefinite article can be removed from here and the section title?

I tried rewriting the "Do not do.. to avoid..", wdyt?

s/For production environments, do not use a KIE scanner or `SNAPSHOT` project versions to avoid accidental or unexpected project updates. The KIE scanner is intended for development environments that use `SNAPSHOT` project versions.
/To avoid accidental or unexpected project updates, do not use KIE scanner or `SNAPSHOT` project versions in production environments. The KIE scanner is intended for development environments that use `SNAPSHOT` project versions.

sterobin replied:
Well, it's considered a tool/service of which you can use multiple. I went back and forth on article usage, and I think "a KIE scanner" is clearest because it communicates that you set it up for a specified container or containers. Thus you can have multiple KIE scanners polling differing KIE containers if configured separately. The other ways of describing it just sounded awkward to me, imo.

Regarding the note. I tried several different structures, including what you propose. Normally I would definitely go with your suggestion because it's almost always best to start with the purpose/action so that the user knows whether he/she cares or not ("To do x, ..."  - "I don't care about that, so I'm not going to read", or "Ooh, yes, what about X." :) )  However, in this case, the emphasis and attention-grabber needs to be about production environments, not avoiding unexpected updates. Those same unexpected updates are likewise a possibility for dev environments too, but in that case, users don't care as much. So I think I still prefer how it stands.

csherrar replied:
Thanks for clarifying that for me. I agree.

csherrar commented on doc-content/shared-kie-docs/src/main/asciidoc/KIE/BuildDeployUtilizeAndRun/kie-scanner-proc.adoc:
```diff
@@ -0,0 +1,83 @@
+For production environments, do not use a KIE scanner or `SNAPSHOT` project versions to avoid accidental or unexpected project updates. The KIE scanner is intended for development environments that use `SNAPSHOT` project versions.
+
+ifdef::PAM,JBPM[]
+Avoid using a KIE scanner with business processes. Using a KIE scanner with processes can lead to unforeseen updates that can then cause errors in long-running processes when changes are not compatible with running process instances.
```
s/Avoid using a/Do not use

sterobin replied:
Well, it's tough because it's a spectrum of discouragement :) It's a concrete NO for using this in production, so I've said "do not ...". For processes though, it's a slightly softer discouragement because users can still knowingly choose to use the scanner with  processes if it's a dev environment and they're aware of the side effects, but just don't mind. So "do not" is too strong for that, imo.

csherrar replied:
I see what you mean, that makes sense.

sterobin commented:
ok to test
